drivers/pwm/pwm_nrfx.c

```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
+		pdata->config.top_value = period_cycles;
+		pdata->prescaler_divisor = 1;
+		pdata->period_cycles = period_cycles;
+	}
+
+	/* Write new PRESCALER and COUNTERTOP to PWM instance */
+	nrf_pwm_configure(pInstance,
+		pdata->config.base_clock,
+		pdata->config.count_mode,
+		pdata->config.top_value);
+
+	return 0;
+}
+
+static u8_t pwm_channel_map(const struct pwm_nrfx_data *pdata, u32_t pwm)
+{
+	u8_t i;
+
+	/* Find pin, return channel number */
+	for (i = 0; i < NRF_PWM_CHANNEL_COUNT; i++) {
+		if (pdata->config.output_pins[i] != NRFX_PWM_PIN_NOT_USED
+		&&
+		(pwm == (pdata->config.output_pins[i] & NRFX_PWM_PIN_MASK))) {
+			return i;
+		}
+	}
+
+	/* Return NRF_PWM_CHANNEL_COUNT to show that PWM pin was not found. */
+	return NRF_PWM_CHANNEL_COUNT;
+}
+
+u16_t get_channel_pulse_cycle(u16_t current)
+{
+	return (current & NRFX_PWM_COMPARE_MASK);
+}
+
+u16_t get_pin_number(u16_t config_output_pin)
+{
+	return (config_output_pin & NRFX_PWM_PIN_MASK);
+}
+
+bool get_pin_inversion_state(u16_t config_output_pin)
+{
+	return (config_output_pin & NRFX_PWM_PIN_INVERTED_MASK);
+}
+
+bool pwm_channel_active(const struct pwm_nrfx_data *pdata)
+{
+	u8_t channel;
+
+	for (channel = 0; channel < NRF_PWM_CHANNEL_COUNT; channel++) {
+		u16_t channel_pulse_cycle =
+			get_channel_pulse_cycle(pdata->current[channel]);
+
+		if (channel_pulse_cycle > 0 &&
+			channel_pulse_cycle < pdata->config.top_value) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
 static int pwm_nrfx_pin_set(struct device *dev, u32_t pwm,
 			    u32_t period_cycles, u32_t pulse_cycles)
 {
+	/* We assume here that period_cycles will always be 16NHz
+	 * peripheral clock. Since pwm_nrfx_get_cycles_per_sec() function might
+	 * be removed, see ISSUE #6958.
+	 * TODO: Remove this comment when issue has been resolved.
+	 */
 	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
 	struct pwm_nrfx_data *pdata = dev->driver_data;
+	NRF_PWM_Type *pwm_instance = pconfig->pwm.p_registers;
+	u8_t channel;
+	u32_t ret;
 
-	if (pwm >= NRF_PWM_CHANNEL_COUNT) {
+	/* Check if PWM pin is one of the predefiend DTS config pins.
+	 * Return its array index (channel number),
+	 * or NRF_PWM_CHANNEL_COUNT if not initialized through DTS.
+	 */
+	channel = pwm_channel_map(pdata, pwm);
+	if (channel == NRF_PWM_CHANNEL_COUNT) {
+		LOG_ERR("PWM pin %d not enabled through DTS configuration.",
+			pwm);
 		return -EINVAL;
 	}
 
-	if (period_cycles != pdata->top_value) {
-		if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
-			return -EINVAL;
-		}
-
-		pdata->top_value = period_cycles;
-		nrf_pwm_configure(pconfig->pwm.p_registers,
-				  pconfig->config.base_clock,
-				  pconfig->config.count_mode,
-				  pdata->top_value);
+	/* Check if period_cycle is either matching currently used, or
+	 * possible to use with our prescaler options.
+	 */
+	ret = pwm_period_check_and_set(pdata, pwm_instance, pwm, period_cycles);
+	if (ret) {
+		LOG_ERR("Incompatible period %d, already in use: %d",
```

Comment:
This message will be shown also when the instance is not in use but there is no suitable prescaler available. In such case "already in use: x" will be only misleading.

Reply:
Agreed. I will remove the already in use comment.